vmm/src/serial_manager.rs

```diff
@@ -0,0 +1,245 @@
+    ReadInput(#[source] io::Error),
+
+    /// Cannot queue input to the serial device.
+    #[error("Error queuing input to the serial device: {0}")]
+    QueueInput(#[source] vmm_sys_util::errno::Error),
+
+    /// Cannot make the file descriptor non-blocking.
+    #[error("Error making input file descriptor non-blocking: {0}")]
+    SetNonBlocking(#[source] io::Error),
+
+    /// Cannot create EventFd.
+    #[error("Error creating EventFd: {0}")]
+    EventFd(#[source] io::Error),
+
+    /// Cannot spawn SerialManager thread.
+    #[error("Error spawning SerialManager thread: {0}")]
+    SpawnSerialManager(#[source] io::Error),
+}
+pub type Result<T> = result::Result<T, Error>;
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+#[repr(u64)]
+pub enum EpollDispatch {
+    File = 0,
+    Kill = 1,
+    Unknown,
+}
+
+impl From<u64> for EpollDispatch {
+    fn from(v: u64) -> Self {
+        use EpollDispatch::*;
+        match v {
+            0 => File,
+            1 => Kill,
+            _ => Unknown,
+        }
+    }
+}
+
+pub struct SerialManager {
+    #[cfg(target_arch = "x86_64")]
+    serial: Arc<Mutex<Serial>>,
+    #[cfg(target_arch = "aarch64")]
+    serial: Arc<Mutex<Pl011>>,
+    epoll_file: File,
+    in_file: File,
+    kill_evt: EventFd,
+    handle: Option<thread::JoinHandle<()>>,
+}
+
+impl SerialManager {
+    pub fn new(
+        #[cfg(target_arch = "x86_64")] serial: Arc<Mutex<Serial>>,
+        #[cfg(target_arch = "aarch64")] serial: Arc<Mutex<Pl011>>,
+        pty_pair: Option<Arc<Mutex<PtyPair>>>,
+        mode: ConsoleOutputMode,
+    ) -> Result<Option<Self>> {
+        let in_file = match mode {
+            ConsoleOutputMode::Pty => {
+                if let Some(pty_pair) = pty_pair {
+                    pty_pair
+                        .lock()
+                        .unwrap()
+                        .main
+                        .try_clone()
+                        .map_err(Error::FileClone)?
+                } else {
+                    return Ok(None);
+                }
+            }
+            ConsoleOutputMode::Tty => {
+                // If running on an interactive TTY then accept input
+                if unsafe { libc::isatty(libc::STDIN_FILENO) == 1 } {
+                    let stdin_clone = unsafe { File::from_raw_fd(libc::dup(libc::STDIN_FILENO)) };
+                    let ret = unsafe {
+                        let mut flags = libc::fcntl(stdin_clone.as_raw_fd(), libc::F_GETFL);
+                        flags |= libc::O_NONBLOCK;
+                        libc::fcntl(stdin_clone.as_raw_fd(), libc::F_SETFL, flags)
+                    };
+
+                    if ret < 0 {
+                        return Err(Error::SetNonBlocking(std::io::Error::last_os_error()));
+                    }
+
+                    stdin_clone
+                } else {
+                    return Ok(None);
+                }
+            }
+            _ => return Ok(None),
```

Comment:
Same here, the caller shouldn't try to create a new SerialManager when the mode is not appropriate, therefore we should throw an error here as well.

Reply:
This is kinda non-obvious but the `SerialManager` is only created (at the moment) where the stdin is a tty. Imagine a situation where you run cloud-hypervisor with --serial tty but have already closed stdin (e.g. like `nohup`) and stdout is redirected to file.

Reply:
The check on the `isatty()` or on the `PtyPair` could be done in the DeviceManager. Or maybe this `::new()` function just needs a comment.

Reply:
@sboeuf For the line 123 return, similarly to the above, I could have made the return an error as the code calling this in `device_manager` checks specifically for the `mode` before calling. I decided to `return Ok(None)` though as that check could reasonably be removed and not creating a `SerialManager` instead of returning an `Error` seemed correct as a `Null` mode would have been a potentially valid case.

Just a note, the different checks needed given the combination of options accepted (`serial_pty` is `Some` and the `mode` is `Pty` `||` `mode` is `Tty` and stdin is as well) I felt were ugly putting into `device_manager`. `device_manager`'s calling of the `SerialManager::new` and `SerialManager::start_thread` are already a pretty ugly bit of code anyway =/. I mostly felt this was my least worst option after poking it a bit but I'm happy to make changes.

Reply:
Yeah I understand. The whole machine state is quite complex and if you feel that's the best we can do, let's go with that!